core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceConcurrencyTest.java

```diff
@@ -0,0 +1,331 @@
+import java.security.KeyPair;
+import java.sql.Date;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class CloudSqlInstanceConcurrencyTest {
+
+  private static final Logger logger =
+      Logger.getLogger(CloudSqlInstanceConcurrencyTest.class.getName());
+
+  private static class TestDataSupplier implements InstanceDataSupplier {
+
+    private volatile boolean flaky;
+
+    private final AtomicInteger counter = new AtomicInteger();
+    private final AtomicInteger successCounter = new AtomicInteger();
+    private final InstanceData response =
+        new InstanceData(
+            new Metadata(
+                ImmutableMap.of(
+                    "PUBLIC", "10.1.2.3",
+                    "PRIVATE", "10.10.10.10",
+                    "PSC", "abcde.12345.us-central1.sql.goog"),
+                null),
+            new SslData(null, null, null),
+            Date.from(Instant.now().plus(1, ChronoUnit.HOURS)));
+
+    private TestDataSupplier(boolean flaky) {
+      this.flaky = flaky;
+    }
+
+    @Override
+    public InstanceData getInstanceData(
+        CloudSqlInstanceName instanceName,
+        AccessTokenSupplier accessTokenSupplier,
+        AuthType authType,
+        ListeningScheduledExecutorService executor,
+        ListenableFuture<KeyPair> keyPair)
+        throws ExecutionException, InterruptedException {
+
+      // This method mimics the behavior of SqlAdminApiFetcher under flaky network conditions.
+      // It schedules a future on the executor to produces the result InstanceData.
+      // When `this.flaky` is set, every other call to getInstanceData()
+      // will pause the thread for an extra 500ms and then throw an ExecutionException,
+      // as if SqlAdminApiFetcher made an API request, and that request took extra time
+      // and then failed.
+      ListenableFuture<InstanceData> f =
+          executor.schedule(
+              () -> {
+                int c = counter.incrementAndGet();
+                if (flaky && c % 2 == 0) {
+                  throw new ExecutionException("Flaky", new Exception());
+                }
+                successCounter.incrementAndGet();
+                return response;
+              },
+              100,
+              TimeUnit.MILLISECONDS);
+
+      return f.get();
+    }
+  }
+
+  private static class TestCredentialFactory implements CredentialFactory, HttpRequestInitializer {
+
+    @Override
+    public HttpRequestInitializer create() {
+      return this;
+    }
+
+    public void initialize(HttpRequest var1) throws IOException {
+      // do nothing
+    }
+  }
+
+  @Test(timeout = 45000)
```

Comment:
Do we still need this?

Reply:
Yes. If there is a deadlock, then the test will only fail with a timeout. So we should leave this in just incase we cause a deadlock again in the future.

Reply:
Do we really want to wait 45 seconds before failing? What's the default value?